### PR TITLE
DG - Fees Calc and Undo/Redo

### PR DESCRIPTION
**Pull request by angkoonhwee**
- [x] Efforts for undo and redo
- [x] Efforts for fees calc

### CODE REVIEW

File: docs/DeveloperGuide.md [hunk 1]
```diff
@@ -149,6 +150,7 @@ The `Model` component
   * While `UniqueTagList` stores all existing distinct tags created, `Tag` objects with the same tag name may not have the same reference. This means that each `Person` object still has its own `Tag` objects and the `Tag` is only added into `UniqueTagList` if it does not exist in the tag list yet.
 * stores the currently 'selected' `Person` objects (e.g., results of a search query) as a separate filtered list which is exposed to outsiders as an unmodifiable `ObservableList<Person>` that can be 'observed', e.g., the `Ui` component can be bound to this list so that it automatically updates when the data in the list change.
 * stores a `UserPref` object that represents the user’s preferences. This is exposed to the outside as a `ReadOnlyUserPref` objects.
+* stores the `LocalDateTime` object of the date and time the `AddressBook` was updated.
```
Comment (Chesterwongz) [inline docs/DeveloperGuide.md:153]:
```suggestion
* stores a`LocalDateTime` object of the date and time the `AddressBook` was last updated.
```

File: docs/DeveloperGuide.md [hunk 2]
```diff
@@ -523,6 +525,43 @@ The redo does the exact opposite (pops from `redoStack`, push to `undoStack`, an
...
+The following activity diagram in Figure I.2.8 summarizes what happens inside the `UndoRedoStack` when a user executes a new command:
+
+![UndoRedoStackActivityDiagram](images/UndoRedoStackActivityDiagram.png)
+
+*Figure I.2.8: Activity Diagram after a `Command` is executed.*
+
+In Figure I.2.8, after execution, any `Command` that is **not** an `UndoableCommand` will not change the `UndoRedoStack`.
+If the `Command` is an `UndoableCommand`, it will be pushed to the `undoStack` before the control is returned to the user eventually.
 
 #### Design considerations:
```
Comment (Chesterwongz) [inline docs/DeveloperGuide.md:566]: remove colons
```suggestion
#### Design considerations
```

File: docs/DeveloperGuide.md [hunk 3]
```diff
@@ -614,6 +653,81 @@ The predicate is then used to filter the list of person.
...
+### Fees Management
+
+TAB automates the updating of individual lesson fees after the lesson has ended. The Fees Calculation is facilitated by:
+* `FeesCalculator` implements the `Calculator` interface. It is responsible for calculates the amount to update for each specific lesson. 
+* `LastUpdatedDate` stores the `LocalDateTime` of when the `AddressBook` was last updated. `LastUpdatedDate` is stored in `AddressBook`.
+* `Money` and its subclass facilitates the payment mechanism. Each `Lesson` contains a `LessonRates` field and a `OutstandingFees` field. Both of which extends `Money`.
+
+When the user launches TAB, in `MainApp#initModelManager()`, the `FeesCalculator` would update all the outstanding lesson fees in the model. Figure I.6.1 is a sequence diagram after user launches TAB.
+
+![UpdateFeesSequenceDiagram](images/UpdateFeesSequenceDiagram.png)
```
Comment (Chesterwongz) [inline docs/DeveloperGuide.md:665]: Should be grey

File: docs/DeveloperGuide.md [hunk 4]
```diff
@@ -614,6 +653,81 @@ The predicate is then used to filter the list of person.
     * Pros: It offers greater flexibility for each field predicate to have its own matching behavior.
     * Cons: There is greater probability of bugs introduced if a new field to search is to be added, or if the matching behaviour of all predicates are required to be changed.
 
+### Fees Management
+
+TAB automates the updating of individual lesson fees after the lesson has ended. The Fees Calculation is facilitated by:
+* `FeesCalculator` implements the `Calculator` interface. It is responsible for calculates the amount to update for each specific lesson.
```
Comment (Xiaoyunnn) [inline docs/DeveloperGuide.md:659]:
```suggestion
* `FeesCalculator` implements the `Calculator` interface. It is responsible for calculating the amount to update for each specific lesson. 
```

File: docs/DeveloperGuide.md [hunk 5]
```diff
@@ -614,6 +653,81 @@ The predicate is then used to filter the list of person.
...
+![UpdateFeesSequenceDiagram](images/UpdateFeesSequenceDiagram.png)
+
+*Figure I.6.1: Sequence diagram of Update Fees.*
+
+In `initModelManager()`, after the `model` was built from storage, `FeesCalculator#UpdateAllLessonOutstanding()` is called to update the lesson fees. For each person in `addressBook`, a new `updatedPerson` would be created.
+When creating the new `updatedPerson`, for each lesson that the person has, `updateLessonOutstandingFeesField()` will update all the outstanding fees fields using `lastUpdatedDate` and the current local date time.  
+
+
+<div markdown="span" class="alert alert-info">:information_source: **Note:** 
+FeesCalculator do not refresh the outstanding fees immediately after the lesson ends if TAB is open. Users need to relaunch TAB to see the updated outstanding fees. </div>
```
Comment (Chesterwongz) [inline docs/DeveloperGuide.md:674]:
```suggestion
FeesCalculator does not refresh the outstanding fees immediately after the lesson ends if TAB is open. Users need to relaunch TAB to see the updated outstanding fees. </div>
```

File: docs/DeveloperGuide.md [hunk 6]
```diff
@@ -614,6 +653,81 @@ The predicate is then used to filter the list of person.
     * Pros: It offers greater flexibility for each field predicate to have its own matching behavior.
     * Cons: There is greater probability of bugs introduced if a new field to search is to be added, or if the matching behaviour of all predicates are required to be changed.
 
+### Fees Management
+
+TAB automates the updating of individual lesson fees after the lesson has ended. The Fees Calculation is facilitated by:
+* `FeesCalculator` implements the `Calculator` interface. It is responsible for calculates the amount to update for each specific lesson. 
+* `LastUpdatedDate` stores the `LocalDateTime` of when the `AddressBook` was last updated. `LastUpdatedDate` is stored in `AddressBook`.
+* `Money` and its subclass facilitates the payment mechanism. Each `Lesson` contains a `LessonRates` field and a `OutstandingFees` field. Both of which extends `Money`.
```
Comment (Xiaoyunnn) [inline docs/DeveloperGuide.md:661]:
```suggestion
* `Money` and its subclass facilitates the payment mechanism. Each `Lesson` contains a `LessonRates` field and a `OutstandingFees` field. Both of which extend `Money`.
```

File: docs/DeveloperGuide.md [hunk 7]
```diff
@@ -614,6 +653,81 @@ The predicate is then used to filter the list of person.
...
+
+When the user launches TAB, in `MainApp#initModelManager()`, the `FeesCalculator` would update all the outstanding lesson fees in the model. Figure I.6.1 is a sequence diagram after user launches TAB.
```
Comment (Xiaoyunnn) [inline docs/DeveloperGuide.md:663]:
```suggestion
When the user launches TAB, in `MainApp#initModelManager()`, the `FeesCalculator` would update all the outstanding lesson fees in the model. Figure I.6.1 is a sequence diagram after the user launches TAB.
```

File: docs/DeveloperGuide.md [hunk 8]
```diff
@@ -614,6 +653,81 @@ The predicate is then used to filter the list of person.
...
+
+![UpdateFeesSequenceDiagram](images/UpdateFeesSequenceDiagram.png)
+
+*Figure I.6.1: Sequence diagram of Update Fees.*
+
+In `initModelManager()`, after the `model` was built from storage, `FeesCalculator#UpdateAllLessonOutstanding()` is called to update the lesson fees. For each person in `addressBook`, a new `updatedPerson` would be created.
```
Comment (Xiaoyunnn) [inline docs/DeveloperGuide.md:669]:
```suggestion
In `initModelManager()`, after the `model` has been built from the storage, `FeesCalculator#UpdateAllLessonOutstanding()` is called to update the lesson fees. For each person in `addressBook`, a new `updatedPerson` would be created.
```

File: docs/DeveloperGuide.md (hunk 5 above)
Comment (Xiaoyunnn) [inline docs/DeveloperGuide.md:674]:
```suggestion
FeesCalculator does not update the outstanding fees immediately after a lesson ends when TAB is open. Users need to relaunch TAB to see the updated outstanding fees. </div>
```

File: docs/DeveloperGuide.md [hunk 9]
```diff
@@ -614,6 +653,81 @@ The predicate is then used to filter the list of person.
...
+    If both conditions are true, update the fees by adding the cost of the lesson to the current outstanding fees.
+  * **For Recurring Lessons:** 
+    * Find the `laterStart` by comparing `startDate` of lesson and `lastUpdatedDate`. This is to handle cases where lesson starts after the last time TAB was launched.
+    * Find the `earlierEnd` by comparing `endDate` of lesson and `currentDateTime`. This is to handle cases where lesson ends before the current launching of TAB.
+    * Calculate the number of lessons that have passed between the `laterStart` and `earlierEnd`.
+    * Check if any of the cancelled dates falls between `laterStart` and `earlierEnd`. If true, deduct the number of lessons accordingly.
+    * Check for the cases in which the day of `lastUpdatedDate` and `currentDateTime` falls on the day of the lesson and deduct accordingly.
+* Multiply the number of lessons calculated with the cost per lesson to get the amount to be added to `OutstandingFees` field.
+
+![UpdateFeesActivityDiagram](images/UpdateFeesActivityDiagram.png)
```
Comment (Chesterwongz) [inline docs/DeveloperGuide.md:694]: Could introduce the diagram first, right now i dont know what i'm looking at, why does it start at "copy..."?

File: docs/DeveloperGuide.md [hunk 10]
```diff
@@ -1255,6 +1369,43 @@ Incorrect `remind` command to try:
 
 Expected: Error details will be shown in the status message.
 
+### Fees Calculator
+
+<div markdown="span" class="alert alert-info">:information_source: <b>Note:</b> The test cases below are designed separately and independent of each other. i.e. Test the different test cases separately.</div>
+
+1. Add the recurring lessons and makeup lessons to a student, following the test cases. <br>
+   Prerequisites: The lessons do not clash with each other and other existing lessons. If there were clashes, delete the clashing lesson(s).
+2. Close TAB and open the json file `/data/addressbook.json`.
```
Comment (Xiaoyunnn) [inline docs/DeveloperGuide.md:1378]:
```suggestion
2. Close TAB and open the json file at `./data/addressbook.json`. 
```

File: docs/DeveloperGuide.md [hunk 11]
```diff
@@ -1255,6 +1369,43 @@ Incorrect `remind` command to try:
...
+2. Close TAB and open the json file `/data/addressbook.json`. 
+3. At the bottom of the json file there is `lastUpdated` field with a `DateTime` object. Edit the `DateTime` object according to the test cases.
+
+<div markdown="span" class="alert alert-info">:information_source: <b>Note:</b> Please vary the date and time accordingly (adding or subtracting by an equivalent amount of day(s) and time), 
+depending on the date and time when you are testing.</div>
+
+Suppose the date today is 6 Nov 2021 and current time is 1800 hours,
+* Test case: 
+  1. `ladd 1 recurring/ date/30 Oct 2021 time/1400-1500 rates/25 subject/math`
+  2. Set the `DateTime` in `data/addressbook.json` to `2021-10-31T12:00`
```
Comment (Xiaoyunnn) [inline docs/DeveloperGuide.md:1387]:
```suggestion
  2. Set the `DateTime` in `./data/addressbook.json` to `2021-10-31T12:00`
```

File: docs/DeveloperGuide.md [hunk 12]
```diff
@@ -1255,6 +1369,43 @@ Incorrect `remind` command to try:
...
+  Expected: Outstanding lesson fees for first lesson of the first person is `$25`. The lesson that have passed is:
+    * 6 Nov 2021 1400-1500
+
+* Test case:
+  1. `ladd 1 recurring/ date/30 Oct 2021 time/1400-1500 rates/25 subject/math`
+  2. Set the `DateTime` in `data/addressbook.json` to `2021-10-30T12:00`
```
Comment (Xiaoyunnn) [inline docs/DeveloperGuide.md:1393]:
```suggestion
  2. Set the `DateTime` in `./data/addressbook.json` to `2021-10-30T12:00`
```

File: docs/DeveloperGuide.md [hunk 13]
```diff
@@ -614,6 +653,81 @@ The predicate is then used to filter the list of person.
...
+
+The following is the logic of `FeesCalculator` to decide whether to update the specific lesson:
+* To construct a `FeesCalculator` object, the `lastUpdatedDate` and `currentDateTime`, which is simply the `LocalDateTime.now()`, are required. 
+* `FeesCalculator` would count the number of lessons that have ended between `lastUpdatedDate` and `currentDateTime` at the time of initialization.
+  * **For Makeup Lessons:** 
+    * Use `Date#getLocalDate()` to get the date of the lesson and `TimeRange#getEnd()` to get the end time of the lesson. Then check for these conditions:
+      * Lesson end date and time is after `lastUpdatedDate`.
+      * Lesson end date and time is before `currentDateTime`.
+      
+    If both conditions are true, update the fees by adding the cost of the lesson to the current outstanding fees.
```
Comment (Xiaoyunnn) [inline docs/DeveloperGuide.md:685]:
```suggestion
    If both conditions are true, update the fees by adding the cost of the lesson (as calculated by multiplying the hourly rates and the duration of the lesson) to the current outstanding fees.
```

File: docs/DeveloperGuide.md [hunk 14]
```diff
@@ -1255,6 +1369,43 @@ Incorrect `remind` command to try:
...
+* Test case:
+  1. `ladd 1 recurring/ date/30 Oct 2021 time/1400-1500 rates/25 subject/math`
+  2. Set the `DateTime` in `data/addressbook.json` to `2021-10-30T12:00`
+  Expected: Outstanding lesson fees for the first lesson of the first person is `$50`. The lessons that have passed are:
+     * 30 Oct 2021 1400-1500
+     * 6 Nov 2021 1400-1500
+
+* Test case:
+  1. `ladd 1 recurring/ date/3 Nov 2021 time/1400-1500 rates/25 subject/math` 
+  2. Set the `DateTime` in `data/addressbook.json` to `2021-10-20T00:00`
```
Comment (Xiaoyunnn) [inline docs/DeveloperGuide.md:1400]:
```suggestion
  2. Set the `DateTime` in `./data/addressbook.json` to `2021-10-20T00:00`
```

File: docs/DeveloperGuide.md [hunk 15]
```diff
@@ -1255,6 +1369,43 @@ Incorrect `remind` command to try:
...
+3. At the bottom of the json file there is `lastUpdated` field with a `DateTime` object. Edit the `DateTime` object according to the test cases.
+
+<div markdown="span" class="alert alert-info">:information_source: <b>Note:</b> Please vary the date and time accordingly (adding or subtracting by an equivalent amount of day(s) and time), 
+depending on the date and time when you are testing.</div>
+
+Suppose the date today is 6 Nov 2021 and current time is 1800 hours,
+* Test case: 
+  1. `ladd 1 recurring/ date/30 Oct 2021 time/1400-1500 rates/25 subject/math`
+  2. Set the `DateTime` in `data/addressbook.json` to `2021-10-31T12:00`
+  Expected: Outstanding lesson fees for first lesson of the first person is `$25`. The lesson that have passed is:
```
Comment (Xiaoyunnn) [inline docs/DeveloperGuide.md:1388]:
```suggestion
  Expected: Outstanding lesson fees for the first lesson of the first student is `$25`. The lesson that has passed is:
```

File: docs/DeveloperGuide.md [hunk 16]
```diff
@@ -1255,6 +1369,43 @@ Incorrect `remind` command to try:
...
+    * 6 Nov 2021 1400-1500
+
+* Test case:
+  1. `ladd 1 recurring/ date/30 Oct 2021 time/1400-1500 rates/25 subject/math`
+  2. Set the `DateTime` in `data/addressbook.json` to `2021-10-30T12:00`
+  Expected: Outstanding lesson fees for the first lesson of the first person is `$50`. The lessons that have passed are:
```
Comment (Xiaoyunnn) [inline docs/DeveloperGuide.md:1394]:
```suggestion
  Expected: Outstanding lesson fees for the first lesson of the first student is `$50`. The lessons that have passed are:
```

File: docs/DeveloperGuide.md [hunk 17]
```diff
@@ -1255,6 +1369,43 @@ Incorrect `remind` command to try:
...
+     * 30 Oct 2021 1400-1500
+     * 6 Nov 2021 1400-1500
+
+* Test case:
+  1. `ladd 1 recurring/ date/3 Nov 2021 time/1400-1500 rates/25 subject/math` 
+  2. Set the `DateTime` in `data/addressbook.json` to `2021-10-20T00:00`
+  Expected: Outstanding lesson fees for the first lesson of the first person is `$25`. The lesson that have passed is:
```
Comment (Xiaoyunnn) [inline docs/DeveloperGuide.md:1401]:
```suggestion
  Expected: Outstanding lesson fees for the first lesson of the first student is `$25`. The lesson that has passed is:
```

File: docs/DeveloperGuide.md [hunk 18]
```diff
@@ -1255,6 +1369,43 @@ Incorrect `remind` command to try:
...
+     * 3 Nov 2021 1400-1500
+
+* Test case:
+  1. `ladd 1 date/6 Nov 2021 time/2000-2130 rates/25 subject/math`
+  2. Set the `DateTime` in `data/addressbook.json` to `2021-11-06T18:00`
```
Comment (Xiaoyunnn) [inline docs/DeveloperGuide.md:1406]:
```suggestion
  2. Set the `DateTime` in `./data/addressbook.json` to `2021-11-06T18:00`
```

File: docs/DeveloperGuide.md [hunk 19]
```diff
@@ -1255,6 +1369,43 @@ Incorrect `remind` command to try:
...
+  Expected: Outstanding lesson fees for first lesson of the first person is `$0`. No lessons have passed.
```
Comment (Xiaoyunnn) [inline docs/DeveloperGuide.md:1407]:
```suggestion
  Expected: Outstanding lesson fees for the first lesson of the first student is `$0`. No lessons have passed.
```

File: docs/DeveloperGuide.md [hunk 20]
```diff
@@ -1255,6 +1369,43 @@ Incorrect `remind` command to try:
 
 Expected: Error details will be shown in the status message.
 
+### Fees Calculator
+
+<div markdown="span" class="alert alert-info">:information_source: <b>Note:</b> The test cases below are designed separately and independent of each other. i.e. Test the different test cases separately.</div>
+
+1. Add the recurring lessons and makeup lessons to a student, following the test cases. <br>
+   Prerequisites: The lessons do not clash with each other and other existing lessons. If there were clashes, delete the clashing lesson(s).
```
Comment (Xiaoyunnn) [inline docs/DeveloperGuide.md:1377]:
```suggestion
   Prerequisites: The lessons do not clash with each other and other existing lessons. If there were clashes, delete the clashing lesson(s). The student to which the lesson is to be added has no other lessons.
```
Adding this line cus ur expected result specifies that the lesson is the first lesson for the student

**Review by Chesterwongz — APPROVED** [review]

**Pull request merged**